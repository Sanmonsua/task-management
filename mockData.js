export const mockCategory = {
	id : '1267494525246',
	name : 'Work',
	tasks : [
		{
			id : 1,
			name: 'Make changes to the new site design',
			date : '15 sept 2020',
			done : false,
		},
		{
			id : 2,
			name: 'Try new icon set',
			date : '11 sept 2020',
			done : true,
		},
	],
	color : '#a362ea',
}

export const mockUser = {
	uid : 'admin123',
}

export const mockTask = {
	categoryId : '1267494525246',
	id : '1600662268405',
	name: 'Make changes to the new site design',
	date : '1600662268405',
	done : false,
}

export const mockCategories = {
	'1267494525246' : {
		id : '1267494525246',
		name : 'Work',
		tasks : [
			{
				id : 1,
				name: 'Make changes to the new site design',
				date : '15 sept 2020',
				done : false,
			},
			{
				id : 2,
				name: 'Try new icon set',
				date : '11 sept 2020',
				done : true,
			},
		],
		color : '#a362ea',
	}
}