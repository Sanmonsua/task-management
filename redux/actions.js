import { ADD_CATEGORY, ADD_TASK, TOGGLE_TASK, EDIT_TASK, SELECT_CATEGORY, FETCH_CATEGORIES } from './actionTypes'
import {firebaseApp} from '../firebase'

let newCategoryId = 0 
let newTaskId = 0
// action creators
export const addCategory = newCategory => {
  newCategoryId ++
  return ({
      type: ADD_CATEGORY,
      payload: {
        ... newCategory, 
        id : `${newCategoryId}`,
        tasks : [],
      },
    }
  )
}

export const addTask = newTask => {
  newTaskId ++
  return ({
      type: ADD_TASK,
      payload: {
        ...newTask,
        id : `${newTaskId}`,
        done : false,
      },
    }
  )
}

export const editTask = task => {
  return ({
    type : EDIT_TASK,
    payload : task,
  })
}

export const toggleTask = task => ({
  type: TOGGLE_TASK,
  payload: task,
})

export const selectCategory = category =>({
  type: SELECT_CATEGORY,
  payload: category,
})

export const fetchCategories = () => dispatch => {
  
  firebaseApp.database().ref('/categories').on('value', snap =>{
    
    dispatch({
      type: FETCH_CATEGORIES,
      payload: {categories : snap.val()},
    })
  })
}