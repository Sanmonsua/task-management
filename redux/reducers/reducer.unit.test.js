import reducer from './index'
import * as actions from '../actions'
import { mockUser, mockCategories, mockTask, mockCategory } from '../../mockData'


const DEFAULT_STATE = {
	user : {},
	categories : {
		allIds : ['1267494525246'],
		byIds : mockCategories,
		selectedId : null, 
	}
}


describe('User reducer', () => {
	it('signIn action', () => {
		expect(reducer(DEFAULT_STATE, actions.signIn(mockUser))).toMatchSnapshot()
	})
})

describe('Categories reducer', () => {

    Date.now = jest.fn(() => 1482363367071)

	it('fetchCategories action', () => {
		const firebaseTest = jest.genMockFromModule('firebase')

		const snapshot = { 
			val: () => mockCategories, 
			exportVal: () => mockCategories, 
			exists: jest.fn(() => true) 
		}

		firebaseTest.database = jest.fn().mockReturnValue({
			ref: jest.fn().mockReturnThis(),
			on: jest.fn((eventType, callback) => callback(snapshot)),
			update: jest.fn(() => Promise.resolve(snapshot)),
			remove: jest.fn(() => Promise.resolve()),
			once: jest.fn(() => Promise.resolve(snapshot)),
		})
        
		const mockDispatch = jest.fn()
		actions.fetchCategories(mockUser, firebaseTest)(mockDispatch)

		expect(reducer(DEFAULT_STATE, mockDispatch.mock.calls[0][0])).toMatchSnapshot()
	})

	it('AddCategory action', () => {
		expect(reducer(DEFAULT_STATE, actions.addCategory(mockCategory))).toMatchSnapshot()
	})

	it('AddTask action', () => {
		expect(reducer(DEFAULT_STATE, actions.addTask(mockTask))).toMatchSnapshot()
    })
    
	it('Edit category action', () =>{
		expect(reducer(DEFAULT_STATE, actions.editCategory(mockCategory))).toMatchSnapshot()	
	})


})
