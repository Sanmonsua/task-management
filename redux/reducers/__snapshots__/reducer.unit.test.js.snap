// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories reducer AddCategory action 1`] = `
Object {
  "categories": Object {
    "allIds": Array [
      "1267494525246",
      undefined,
    ],
    "byIds": Object {
      "1267494525246": Object {
        "color": "#a362ea",
        "id": "1267494525246",
        "name": "Work",
        "tasks": Array [
          Object {
            "date": "15 sept 2020",
            "done": false,
            "id": 1,
            "name": "Make changes to the new site design",
          },
          Object {
            "date": "11 sept 2020",
            "done": true,
            "id": 2,
            "name": "Try new icon set",
          },
        ],
      },
      "1482363367071": Object {
        "color": "#a362ea",
        "id": "1482363367071",
        "name": "Work",
        "tasks": Array [],
      },
    },
    "selectedId": "1482363367071",
  },
  "user": Object {},
}
`;

exports[`Categories reducer AddTask action 1`] = `
Object {
  "categories": Object {
    "allIds": Array [
      "1267494525246",
    ],
    "byIds": Object {
      "1267494525246": Object {
        "color": "#a362ea",
        "id": "1267494525246",
        "name": "Work",
        "tasks": Array [
          Object {
            "date": "15 sept 2020",
            "done": false,
            "id": 1,
            "name": "Make changes to the new site design",
          },
          Object {
            "date": "11 sept 2020",
            "done": true,
            "id": 2,
            "name": "Try new icon set",
          },
          Object {
            "categoryId": "1267494525246",
            "date": "1600662268405",
            "done": false,
            "id": "1482363367071",
            "name": "Make changes to the new site design",
          },
        ],
      },
    },
    "selectedId": null,
  },
  "user": Object {},
}
`;

exports[`Categories reducer Edit category action 1`] = `
Object {
  "categories": Object {
    "allIds": Array [
      "1267494525246",
    ],
    "byIds": Object {
      "1267494525246": Object {
        "color": "#a362ea",
        "id": "1267494525246",
        "name": "Work",
        "tasks": Array [
          Object {
            "date": "15 sept 2020",
            "done": false,
            "id": 1,
            "name": "Make changes to the new site design",
          },
          Object {
            "date": "11 sept 2020",
            "done": true,
            "id": 2,
            "name": "Try new icon set",
          },
        ],
      },
    },
    "selectedId": null,
  },
  "user": Object {},
}
`;

exports[`Categories reducer fetchCategories action 1`] = `
Object {
  "categories": Object {
    "allIds": Array [
      "1267494525246",
    ],
    "byIds": Object {
      "1267494525246": Object {
        "color": "#a362ea",
        "id": "1267494525246",
        "name": "Work",
        "tasks": Array [
          Object {
            "date": "15 sept 2020",
            "done": false,
            "id": 1,
            "name": "Make changes to the new site design",
          },
          Object {
            "date": "11 sept 2020",
            "done": true,
            "id": 2,
            "name": "Try new icon set",
          },
        ],
      },
    },
    "selectedId": null,
  },
  "user": Object {},
}
`;

exports[`User reducer signIn action 1`] = `
Object {
  "categories": Object {
    "allIds": Array [
      "1267494525246",
    ],
    "byIds": Object {
      "1267494525246": Object {
        "color": "#a362ea",
        "id": "1267494525246",
        "name": "Work",
        "tasks": Array [
          Object {
            "date": "15 sept 2020",
            "done": false,
            "id": 1,
            "name": "Make changes to the new site design",
          },
          Object {
            "date": "11 sept 2020",
            "done": true,
            "id": 2,
            "name": "Try new icon set",
          },
        ],
      },
    },
    "selectedId": null,
  },
  "user": Object {
    "isAuthenticated": true,
  },
}
`;
